### ISSUE TITLE
MQTT transport doesn't seem to support TLS connections to broker

### ISSUE THREAD

**Issue by oej** [meetecho/janus-gateway#1048]
It seems like the MQTT transport doesn't support connecting to a TLS enabled broker. We can connect to a TCP connection just fine, but with TLS enabled janus fails to connect.

We add a URL like "SSL://server.domain.tld:8883" and it fails miserably. I think the code may need SSL_options and some other stuff in order to support this kind of connection.

Solution: Either document that it's not supported, or document how it's support and fix the code :-)

This may be a feature request, but if it is, the bug is the documentation.

**Comment by lminiero** [comment]
You're probably right: the MQTT was contributed so I'm less familiar with it than I am with the other plugins, but it makes sense that SSL options may be needed, as is the case for the RabbitMQ transport. I'll have a look at the library API as soon as I have some time.

**Comment by lminiero** [comment]
Pinging @manifest, the MQTT transport plugin author, just in case he can nudge us in the right direction.

**Comment by oej** [comment]
As always I would prefer to use TLS. And as always, I don't like that the URL starts with "SSL:" but that is a problem in the MQTT library I need to report elsewhere :-)

**Comment by lminiero** [comment]
@oej can you check the above PR? I implemented it by simply looking at the docs but don't have time to test it unfortunately. Please let me know if that fixes it for you: there are new properties in the config to enable SSL for the connection.

**Comment by lminiero** [comment]
Ping! :slightly_smiling_face:

**Comment by manifest** [comment]
Hi there! In our case Janus communicates with MQTT broker in a safe internal network. In that case using SSL looks kind of overkill. For most cases that should be this way, I believe. From the first sight PR looks ok, I could check it more carefully later.

**Comment by lminiero** [comment]
@manifest thx!

**Comment by oej** [comment]
Wakeup. Will check.

@manifest "Safe internal network" is usually a bad assumption. Being able to run TLS makes it possible to do even more things.

**Comment by manifest** [comment]
It required some more changes, so I've made another [PR](https://github.com/meetecho/janus-gateway/pull/1064). Note that to support ssl/tls we need to link `paho-mqtt3as` library. Should not be a big deal :-)

**Comment by manifest** [comment]
There is [gist](https://gist.github.com/manifest/aa2cfb19b596669b710bbc0f8885d64f). It coud be used to set up development environment, build and start playing with the plugin.

**Comment by lminiero** [comment]
@manifest thanks! I guess it's indeed better you're tackling this now... thanks for the document too, it will definitely be useful to implementers. From a quick glance I don't understand what you added to my PR, though, can you summarize what was missing? If it's just a few changes, it might be easier to create a new PR that is based on my branch, instead of replicating the same changes in a new PR. I'll be busy all week so I'll probably be slow at actually reviewing all this, but I hope I'll find some time to do that soon.

Just as a quick question, when you say we need `paho-mqtt3as`, is this a replacement of the previous dependency, or an addition to it? If a replacement, does this mean we can ONLY do the secure channel? I'd rather keep both options available, as we have for all transports, if possible.

**Comment by manifest** [comment]
@lminiero  I've chery-picked your commit and then squashed with my changes just to make history cleaner. Changes:
- Fixed typos in config key names, like this one `janus_config_get_item_drilldown(config, "general", "ctx->ssl_cert")`
- Set `verify_peer` to `FALSE` by default
- Added a check if `certfile` is presented in config (it's required)
- Added a check if `certfile` and `keyfile` are both presented in config or neither of them (they optional and make sense only when both are presented)

As for the library. `paho-mqtt3as` replaces `paho-mqtt3a` in a sense that we could use both secure and insecure connections. That's why I think it would be ok to just replace one with another. And we always can add both options later as a different feature.

**Comment by lminiero** [comment]
Thanks for clarifying! I'll close my old PR then :+1:

**Comment by lminiero** [comment]
Closing as we merged #1064.
